doc/design.md

```diff
@@ -0,0 +1,113 @@
+# Design Principles
+
+## Security
+
+### No high privilege - no root
+
+For historical reasons smtpd servers required root privilege to run. This
+necessity was explained by several things:
+
+- the ability to use TCP port 25 on Unix-like systems requires root privilge.
+  (on modern Linux at least `CAP_NET_BIND_SERVICE` capability is requred)
+
+- the ability to ignore DAC and write mail to user's directory requries root
+  (`CAP_CHOWN` and `CAP_DAC_OVERRIDE`)
+
+- amusingly, the ability to drop privilege requies root too. 
+  (to employ privilege separation techniques, e.g. spawn child
+  processes and set their UID and GID to a non-privileged user root or `CAP_SETUID`
+  and `CAP_SETGID` capabilities are needed)
+
+
+But all of the above are rudiments of the past.
+
+> It is no longer a catastrophe if an unprivileged process binds to transport
+> layer ports less than 1024. Everyone should consider reading and writing the
+> network medium as unlimited due to hardware no longer costing a million
+> dollars, regardless of what an operating system does.
+
+[Rise and Fall of the Operating System](http://www.fixup.fi/misc/usenix-login-2015/login_oct15_02_kantee.pdf)
+
+The very action of opening TCP port 25 can be delegated to a very small
+privileged program or OS service manager. OS service manager runs with root
+privilege anyway. SMTP server can accept the passed descriptor and use it
+without ever having to escalate privilege.
+
+Nowadays people are not logging in to the mail server via ssh to check mail.
+Modern mail servers are running MDAs like Dovecot to present an IMAP interface
+to the user. Very likely the mail itself is stored in a virtual mailboxes, owned 
+by one user (usually `vmail`)
+
+This practially obsoletes the necessity to support mbox or Maildir in SMTP
+server. Accepted mail should just be handed over to MDA via LMTP.
```

Comment:
It might make sense to also allow piping the email to an executable, this way people who really want to ssh to their server to read their email could do it themselves by doing something like using `sudo` there (pending on whether we set something like `PR_SET_NO_NEW_PRIVS` and how easy it would be to run a sudo-able binary from the receiver)

Reply:
sure